// sheweny/build/structures/Button.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Button = void 0;
const discord_js_1 = require("discord.js");
const _1 = require(".");
/**
 * Represents an Button structure
 * @extends {BaseStructure}
 */
class Button extends _1.BaseStructure {
    /**
     * Constructor for build a Button
     * @param {ShewenyClient} client Client framework
     * @param {string[] | RegExp[]} customId Custom id for one or more buttons
     */
    constructor(client, customId, data) {
        super(client);
        this.cooldown = data?.cooldown || client.managers.buttons?.default?.cooldown;
        this.customId = customId;
        this.manager = this.client.managers.buttons;
    }
    /**
     * Unregister a button from collections
     * @returns {boolean}
     */
    unregister() {
        this.client.collections.buttons?.delete(this.customId);
        delete require.cache[require.resolve(this.path)];
        return true;
    }
    /**
     * Reload a button
     * @returns {Promise<Collection<string[] | RegExp[], Button> | null>}
     */
    async reload() {
        if (this.path) {
            this.unregister();
            return this.register();
        }
        return null;
    }
    /**
     * Register a button in collections
     * @returns {Collection<string[] | RegExp[], Button>}
     */
    async register() {
        const Button = (await Promise.resolve().then(() => require(this.path))).default;
        const btn = new Button(this.client);
        return this.client.collections.buttons
            ? this.client.collections.buttons.set(btn.customId, btn)
            : new discord_js_1.Collection().set(btn.customId, btn);
    }
}
exports.Button = Button;
