// sheweny/build/structures/Modal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Modal = void 0;
const discord_js_1 = require("discord.js");
const BaseStructure_js_1 = require("./BaseStructure.js");
/**
 * Represents an Modal structure
 * @extends {BaseStructure}
 */
class Modal extends BaseStructure_js_1.BaseStructure {
    /**
     * Constructor to build a Modal
     * @param {ShewenyClient} client Client framework
     * @param {string[] | RegExp[]} customId Custom id for one or more modals
     */
    constructor(client, customId, options) {
        super(client);
        this.cooldown = (options?.cooldown || client.managers.buttons?.default?.cooldown) ?? 0;
        this.customId = customId;
    }
    /**
     * Register a modal in collections
     * @returns {Collection<string[] | RegExp[], Modal>}
     */
    async register() {
        const ModalImport = (await Promise.resolve().then(() => require(this.path))).default;
        const mod = new ModalImport(this.client);
        return this.client.collections.modals
            ? this.client.collections.modals.set(mod.customId, mod)
            : new discord_js_1.Collection().set(mod.customId, mod);
    }
    /**
     * Reload a modal
     * @returns {Promise<Collection<string[] | RegExp[], Modal> | ShewenyError>}
     */
    async reload() {
        this.unregister();
        return this.register();
    }
    /**
     * Unregister a modal from collections
     * @returns {boolean}
     */
    unregister() {
        this.client.collections.modals?.delete(this.customId);
        if (!this.path)
            return false;
        delete require.cache[require.resolve(this.path)];
        return true;
    }
}
exports.Modal = Modal;
