// sheweny/build/structures/SelectMenu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectMenu = void 0;
const discord_js_1 = require("discord.js");
const _1 = require(".");
/**
 * Represents an Select Menu structure
 * @extends {BaseStructure}
 */
class SelectMenu extends _1.BaseStructure {
    /**
     * Constructor for build a Select Menu
     * @param {ShewenyClient} client Client framework
     * @param {string[] | RegExp[]} customId Custom id for one or more select menus
     */
    constructor(client, customId, data) {
        super(client);
        this.customId = customId;
        this.cooldown = data?.cooldown || client.managers.selectMenus?.default?.cooldown;
        this.manager = this.client.managers.selectMenus;
    }
    /**
     * Unregister a select menu from collections
     * @returns {boolean}
     */
    unregister() {
        this.client.collections.selectMenus?.delete(this.customId);
        delete require.cache[require.resolve(this.path)];
        return true;
    }
    /**
     * Reload a select menu
     * @returns {Promise<Collection<string[]| RegExp[], SelectMenu> | null>} The select menus collection
     */
    async reload() {
        if (this.path) {
            this.unregister();
            return this.register();
        }
        return null;
    }
    /**
     * Register a select menu in collections
     * @returns {Collection<string[]| RegExp[], SelectMenu>} The select menus collection
     */
    async register() {
        const SelectMenu = (await Promise.resolve().then(() => require(this.path))).default;
        const sm = new SelectMenu(this.client);
        return this.client.collections.selectMenus
            ? this.client.collections.selectMenus.set(sm.customId, sm)
            : new discord_js_1.Collection().set(sm.customId, sm);
    }
}
exports.SelectMenu = SelectMenu;
